[FSDP2] Fixed 2D mismatched grad placements

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #136221
* __->__ #136237

```
CUDA_VISIBLE_DEVICES=2,3,6,7 pytest test/distributed/_composable/test_composability/test_2d_composability.py -k test_train_parity_2d_transformer
```


cc @XilunWu @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Differential Revision: [D62964658](https://our.internmc.facebook.com/intern/diff/D62964658)

Commit:
[FSDP2] Fixed 2D mismatched grad placements

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -695,10 +695,11 @@ def _get_grad_inner_tensor(self, grad: torch.Tensor) -> torch.Tensor:
             if isinstance(grad, AsyncCollectiveTensor):
                 grad = grad.wait()
             assert isinstance(grad, DTensor), f"{type(grad)}"
-            if any(pl.is_partial() for pl in grad.placements):
```
Previously, we changed any partial placements to replicate mainly targeting the case where replicated `RMSNorm.weight` had partial gradients, and we needed to trigger the all-reduce by converting from partial to replicated.

However, for the `pos_embeddings` in our toy `Transformer` class, we have `Shard(0)` placement and still `Partial` grad. We _do not_ want to convert it to `Replicate` but rather to `Shard(0)`.

awgu replied:
I think that size the viewing into reduce-scatter output uses `torch.as_strided`, we were silently handling a larger/replicated gradient for `pos_embeddings.weight`.

awgu commented:
cc: @mori360 do you know if I need to make any changes to CI files to have the added test run in CI? We prefer to run it with 4 GPUs.

mori360 commented:
> do you know if I need to make any changes to CI files to have the added test run in CI? We prefer to run it with 4 GPUs.

The 2d_composability file is tested under multigpu-test.sh, we don't need to add changes in CI file.

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -695,10 +695,11 @@ def _get_grad_inner_tensor(self, grad: torch.Tensor) -> torch.Tensor:
-                placements = [
-                    Replicate() if pl.is_partial() else pl for pl in grad.placements
-                ]
+            placements = self._tp_spec.placements
```
cc: @tianyu-l  on this change just as a heads up

tianyu-l commented on test/distributed/_composable/test_composability/test_2d_composability.py:
```diff
@@ -171,6 +171,62 @@ def _test_train_parity_2d_mlp(
+            for ref_param, (param_name, param) in zip(
+                ref_model.parameters(), model.named_parameters()
+            ):
+                full_grad = param.grad.full_tensor()
```
Should we also assert `param.grad` is sharded here? The test seems not differentiating `Replicate` vs `Shard`

awgu replied:
Sounds good!

awgu replied:
I added a check specifically for `pos_embeddings.weight` and its gradient's TP placement because that was the parameter that exercised the bug. It is hard to assert `param.grad` is sharded generally since we have to case on the parallelize plan (e.g. norm weights are not sharded).

Commit:
Update on "[FSDP2] Fixed 2D mismatched grad placements"


```
CUDA_VISIBLE_DEVICES=2,3,6,7 pytest test/distributed/_composable/test_composability/test_2d_composability.py -k test_train_parity_2d_transformer
```


cc XilunWu H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62964658).

awgu commented:
@pytorchbot merge
